### ISSUE TITLE
Problems using JSON pointers in external $ref

### ISSUE THREAD

**Issue by scalvert** [json-schema-faker/json-schema-faker#644]
If you are reporting a bug, please include the following:

1. the JSON Schema content that fails - preferably using http://json-schema-faker.js.org/ (paste the schema, click `Generate example` and then copy and paste the resulting URL from the browser address bar)

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "Static Analysis Results Format (SARIF) Version 2.1.0-rtm.5 address fragment",
  "id": "https://sarif/address-fragment.json",
  "description": "A dynamic fragment of a SARIF schema.",
  "allOf": [
    {
      "$ref": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/address"
    }
  ]
}
```

2. the result you got

```json
{
  "index": 73052954,
  "absoluteAddress": 53078658,
  "offsetFromParent": 86517330,
  "fullyQualifiedName": "sit quis nostrud laborum id",
  "relativeAddress": -15884468,
  "name": "elit ut do",
  "parentIndex": 32195673,
  "kind": "enim consectetur minim",
  "length": -60095662,
  "properties": {
    "tags": [
      "eu est ad",
      "dolor nisi proident in",
      "nostrud esse ut dolor do",
      "Lorem ea sed Duis sunt",
      "ut cupidatat adipisicing esse"
    ],
    "nostrud_a18": true,
    "autebab": "quis",
    "Excepteur9": true
  }
}
```

3. the result you expected
It passes via the online playground, but fails locally with the following error.

Error: Error while resolving schema (Prop 'definitions' not found in [$schema, title, id, description, allOf] (https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/result))
    at /Users/scalvert/workspace/personal/test-bed/generate-fake-data/node_modules/json-schema-faker/dist/main.cjs.js:3107:13

My schema is a schema that simply contains a JSON pointer to a parent schema (the Microsoft SARIF spec). Since it's an external reference, I'd expected to be able to pass the parent schema in `refs` as the second argument to either `generate` or `resolve`. My understanding (and the way `ajv` works), was that this would allow the pointer to lookup the definitions via the id of the parent schema.

What I don't understand is:
1. Why does the online playground work (I debugged it and it's passing an empty array as the `refs` argument, so I'm not sure how this is working
2. What is the right syntax to be able to use the above schema that references a parent schema.

Any help would be appreciated!

**Comment by scalvert** [comment]
I've looked a bit more into this, and I was slightly incorrect on what I originally posted. Certain definitions, when referenced, work. Certain definitions do not work.

Example:

Working definition:

Schema
```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "Static Analysis Results Format (SARIF) Version 2.1.0-rtm.5 result fragment",
  "id": "https://sarif/address-fragment.json",
  "description": "A dynamic fragment of a SARIF schema.",
  "allOf": [
    {
      "$ref": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/address"
    }
  ]
}
```

Output
```json
{
  "absoluteAddress": 10245663,
  "parentIndex": 31550147,
  "kind": "ut tempor",
  "properties": {
    "tags": [
      "eu reprehenderit irure culpa tempor",
      "velit",
      "eu velit id ipsum nostrud"
    ],
    "eiusmod5": -64128452,
    "cillum_445": 50111525.03080502,
    "eua": false
  },
  "relativeAddress": 83963238,
  "offsetFromParent": 16999010,
  "length": 11026286,
  "fullyQualifiedName": "qui",
  "name": "est",
  "index": 65440184
}
```

Non-working definition:

Schema
```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "Static Analysis Results Format (SARIF) Version 2.1.0-rtm.5 result fragment",
  "id": "https://sarif/result-fragment.json",
  "description": "A dynamic fragment of a SARIF schema.",
  "allOf": [
    {
      "$ref": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/result"
    }
  ]
}
```

Output (in console)
```
json-schema-faker:3139 Uncaught (in promise) Error: Error while resolving schema (Prop 'definitions' not found in [$schema, title, id, description, allOf] (https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/result))
    at json-schema-faker:3139
```

Notice that all that materially changes between the two is the definition referenced in the JSON pointer. It's not clear to me what the difference is here.

**Comment by scalvert** [comment]
It looks as though there's a circular reference in the schema that can't be dereferenced:

```
Uncaught (in promise) Error: Error while resolving schema (Circular $ref pointer found at https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/node/properties/children/items)
```

Any way to resolve this?

**Comment by pateketrueke** [comment]
Can you share the actual code you're using?

**Comment by scalvert** [comment]
It looks like if I change https://github.com/json-schema-faker/json-schema-faker/blob/273d1781d8e2d7594076a9970002832c4623af3c/src/lib/index.js#L141 to use `json-schema-ref-parser`'s `bundle` function instead of `dereference` this works, as it avoids the circular reference issue.

**Comment by scalvert** [comment]
Minimal repro:

```
const jsf = require('json-schema-faker');

let schema = {
      "$schema": "http://json-schema.org/draft-04/schema#",
      "title": "Static Analysis Results Format (SARIF) Version 2.1.0-rtm.5 fragment",
      "id": "https://sarif/result-fragment.json",
      "description": "A dynamic fragment of a SARIF schema.",
      "allOf": [
        {
          "$ref": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/result"
        }
      ]
    };

jsf.resolve(schema).then((data) => {
  console.log(data);
});
```

**Comment by scalvert** [comment]
By calling the `bundle` function vs. `dereference` the error is gone, and the faker data is correctly output. I'd be happy to open a PR to add the option to specify the parser function to use, if you think that'd be useful. I would hate to not use this package as it's amazing, and will save me a massive amount of time.

**Comment by pateketrueke** [comment]
Interesting, even using `.bundle` always almost pass all tests, I wonder if we can switch to it. 🤔

**Comment by scalvert** [comment]
Interesting. What specifically fails? I can push up a branch and we can iterate on that, if that works.

**Comment by scalvert** [comment]
I opened up the PR above if you'd like to discuss. It looks like there's 2 failing tests.
